src/test/java/net/consensys/orion/http/handler/SendRawHandlerTest.java

```diff
@@ -0,0 +1,162 @@
+      final FakePeer fakePeer = new FakePeer(new MockResponse().setBody(digest), memoryKeyStore);
+      // add peer push URL to networkNodes
+      networkNodes.addNode(fakePeer.publicKey, fakePeer.getURL());
+      fakePeers.add(fakePeer);
+    }
+
+    // configureRoutes the binary sendRequest
+    final RequestBody body = RequestBody.create(MediaType.parse(APPLICATION_OCTET_STREAM.httpHeaderValue), toEncrypt);
+    final Box.PublicKey sender = memoryKeyStore.generateKeyPair();
+
+    final String from = encodeBytes(sender.bytesArray());
+
+    final String[] to = fakePeers.stream().map(fp -> encodeBytes(fp.publicKey.bytesArray())).toArray(String[]::new);
+
+    final Request request = new Request.Builder()
+        .post(body)
+        .url(clientBaseUrl + "sendraw")
+        .addHeader("c11n-from", from)
+        .addHeader("c11n-to", String.join(",", to))
+        .addHeader("Content-Type", APPLICATION_OCTET_STREAM.httpHeaderValue)
+        .addHeader("Accept", APPLICATION_OCTET_STREAM.httpHeaderValue)
+        .build();
+
+    // execute request
+    final Response resp = httpClient.newCall(request).execute();
+
+    // ensure we got a 200 OK
+    assertEquals(200, resp.code());
+    // ensure we got the right body
+    assertEquals(digest, resp.body().string());
+  }
+
+  @Test
+  void sendRawApiOnlyWorksOnPrivatePort() throws Exception {
+    // note: this closely mirrors the test "testPropagatedToMultiplePeers",
+    // using the raw version of the API.
+
+    final byte[] toEncrypt = new byte[342];
+    new Random().nextBytes(toEncrypt);
+
+    // encrypt it here to compute digest
+    final EncryptedPayload encryptedPayload = enclave.encrypt(toEncrypt, null, null, null);
+    final String digest = encodeBytes(sha2_512_256(encryptedPayload.cipherText()));
+
+    // create fake peers
+    final FakePeer fakePeer = new FakePeer(new MockResponse().setBody(digest), memoryKeyStore);
+    // add peer push URL to networkNodes
+    networkNodes.addNode(fakePeer.publicKey, fakePeer.getURL());
+
+    // configureRoutes the binary sendRequest
+    final RequestBody body =
+        RequestBody.create(MediaType.parse(HttpContentType.APPLICATION_OCTET_STREAM.httpHeaderValue), toEncrypt);
+    final Box.PublicKey sender = memoryKeyStore.generateKeyPair();
+
+    final String from = encodeBytes(sender.bytesArray());
+
+    final Request request = new Request.Builder()
+        .post(body)
+        .url(nodeBaseUrl + "sendraw")
+        .addHeader("c11n-from", from)
+        .addHeader("c11n-to", encodeBytes(fakePeer.publicKey.bytesArray()))
+        .addHeader("Content-Type", APPLICATION_OCTET_STREAM.httpHeaderValue)
+        .addHeader("Accept", APPLICATION_OCTET_STREAM.httpHeaderValue)
+        .build();
+
+    final Response resp = httpClient.newCall(request).execute();
+
+    // ensure we got a 404 OK
+    assertEquals(404, resp.code());
+  }
+
+  @Test
+  public void sendingWithoutToHeaderSucceeds() throws Exception {
```

Comment:
Is "To" a header, or a field?

Reply:
It is a header. When using /sendraw, we expect two headers:

- `c11n-to` : the `to` address
- `c11n-from`: the `from` address